serf/serf.go

```diff
@@ -1438,6 +1453,33 @@ func (s *Serf) resolveNodeConflict() {
 	}
 }
 
+//handlePrune is a simplified version of Reap() which only runs when the prune
+//flag is set
+func (s *Serf) handlePrune(old []*memberState, m *memberState) []*memberState {
+	// Delete from members
+	delete(s.members, m.Name)
+
+	// Tell the coordinate client the node has gone away and delete
+	// its cached coordinates.
+	if !s.config.DisableCoordinates {
+		s.coordClient.ForgetNode(m.Name)
+
+		s.coordCacheLock.Lock()
+		delete(s.coordCache, m.Name)
+		s.coordCacheLock.Unlock()
```

Comment:
I don't know this code well so I'll go look but is it possible this could race since we `ForgetNode` above and then take lock on cache? It might be OK in that you should only be using this on nodes you know are really dead but I wonder if it's possible for an out-of-order `alive` message from the node to be delivered (or node to come back up) and recreate the client in `coordClient`, updates it's cached coords but then have them removed again here 🤔.

I _guess_ (again not checked the code) the worst case is that we'd not have coordinates until next message from that node so it's not a huge deal probably but just checking. If this is the same pattern we do elsewhere already (e.g. during leave handling) then cool!

Reply:
I think we could totally swap `ForgetNode` and the CacheLock/Delete and it would be easier to read. This function `handlePrune` is a gutted out version of `Reap()`, which I copied the most important parts from - so we do the same pattern there as well. 
As far as out of order alive messages, hopefully the time on the messages will make them easy to refute?

Reply:
@banks the ForgetNode/Lock/delete/Unlock sequence already happens in `reap()`. So if there's a race, it's an old race.

Reply:
It might be clearer if you had both reap() and handlePrune() share the same block of code:
```
func (s *Serf) removeMember(m *memberState, prune bool) {
	// Delete from members
	delete(s.members, m.Name)

	// Tell the coordinate client the node has gone away and delete
	// its cached coordinates.
	if !s.config.DisableCoordinates {
		s.coordClient.ForgetNode(m.Name)

		s.coordCacheLock.Lock()
		delete(s.coordCache, m.Name)
		s.coordCacheLock.Unlock()
	}

	// Send an event along
	if prune {
		s.logger.Printf("[INFO] serf: EventMemberReap (forced): %s", m.Name)
	} else {
		s.logger.Printf("[INFO] serf: EventMemberReap: %s", m.Name)
	}
	if s.config.EventCh != nil {
		s.config.EventCh <- MemberEvent{
			Type:    EventMemberReap,
			Members: []Member{m.Member},
		}
	}
}
```

Then `reap()` can call `s.removeMember(m, false)` and `handlePrune` can call `s.removeMember(m, true)`